### ISSUE TITLE
NilEncoding always dropKey when encoding an Encodable object with optional properties

### ISSUE THREAD

**Issue by amarcadet** [Alamofire/Alamofire#3778]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?

I am using `URLEncodedFormEncoder` with the option `NilEncoding.dropValue` to encode an object that has optional string properties.

## What did you expect to happen?

Assuming the following struct : 

```swift
struct MyStruct: Encodable {
    let p1: String
    let p2: String?
}
```

If I encode myStruct I expect to get the following query `p1=string&p2=`.

```swift
do {
    let myStruct = MyStruct(p1: "string", p2: nil)
    let formEncoder = URLEncodedFormEncoder(nilEncoding: .dropValue)
    let query: String = try formEncoder.encode(myStruct)
    print(query) // prints "p1=string"
} catch {
    print(error)
}
```

## What happened instead?

Got `p1=string`.

## Alamofire Environment

**Alamofire Version:** 5.7.1
**Dependency Manager:** Cocoapods 1.12.1
**Xcode Version:** Xcode 14.3.1
**Swift Version:** 5.7
**Platform(s) Running Alamofire:** iOS
**macOS Version Running Xcode:** Ventura 13.3.1

## Demo Project

Code above is sufficient to reproduce the issue.

## Suggestion

Digging into Alamofire I think the issue might by in `extension _URLEncodedFormEncoder.KeyedContainer: KeyedEncodingContainerProtocol {`.

In fact when the encoder found a nil value it seems to prefer using the `encodeIfPresent` method instead of `encodeNil`.

Using this rather naive implementation I managed to get my Struct encoded properly :

```swift
func encodeIfPresent(_ value: String?, forKey key: Key) throws {
    if let value {
        try encode(value, forKey: key)
    } else {
        guard let nilValue = nilEncoding.encodeNil() else { return }
        try encode(nilValue, forKey: key)
    }
}
```

**Comment by jshier** [comment]
Confirmed, thanks for the report!

Oddly, this isn't the case for `[String: String?]`, so I'll have to figure out why that is. I've replicated your solution but it would only fix it for `String?` values, and I'd rather not have to have an overload for every type.

**Comment by jshier** [comment]
I've fixed this in #3779, feel free to take a look and ensure it works for you before I merge.

**Comment by jshier** [comment]
This has been merged and will go out in the next release. Depending on whatever else I may want to go out, it may be as soon as tomorrow or this weekend.

**Comment by amarcadet** [comment]
Thank you for your quick reply! Awesome 🙏

**Comment by amarcadet** [comment]
Coming back here for an update, do you plan to release a version with the fix soon?

**Comment by jshier** [comment]
I was hoping to release it alongside `WebSocketRequest` but I'm still working on things there. I'll do a release.

**Comment by jshier** [comment]
I've released 5.8.1 with this fix but CocoaPods availability will be delayed pending a backend issue on their side. SPM / Carthage should see it now.
